### ISSUE TITLE
[Shadow Priest] S2MCheck erroneous at high haste values

### ISSUE THREAD

**Issue by pnmueller** [simulationcraft/simc#2780]
While the non-S2M profile keeps scaling linearly with haste, the S2M actually shows a significant DPS decrease when getting more haste after ~10k. Since there's no mechanic that would result in a DPS decrease (as opposed to a lowered DPS increase which can be expected), I strongly suspect the S2MCheck doesn't calculate feasible values for very high haste levels and subsequently S2M is being used too early. This is mainly problematic because it makes it impossible to get proper stat weightings after ~10k haste.

I did a small test and set the raw_haste_pct_100 value to a flat 31 and I'm getting DPS increases from haste at higher haste now. It seems like a small bandaid fix would be to cap the haste used in the formula (e.g. MIN[raw_haste_pct_100;35]). I'm not sure if there's a function for that in Simulationcraft.

**Comment by Twintop** [comment]
I'm already handling this, it's a typo in the nonexecute_actors_pct part.

**Comment by Twintop** [comment]
A little bit more clarity: that formula exists to provide a generic way to tell the sim to pop S2M based around a bunch of variables: raw (unbuffed) haste, what talents you have, what artifact traits you have unlocked, and what percentage of the raid are non-execute specs.

Replacing raw_haste_pct*100 with another value isn't an option, something else in the formula has to change. I'm not sure what version of the APL you're using (or what fight length, or talents, or artifact, or fight type...) but it has gone through massive changes in the last few days in an attempt to make it more responsive to different conditions. I'm certainly not saying it is perfect, but slapping a minimum constant in there doesn't really fix the problem.

**Comment by pnmueller** [comment]
I'm sorry but I feel like I'm not being taken seriously here at all. I don't need an explanation for the formula. If I did, I certainly wouldn't have been able to manually optimize it for certain situations.
The version of the APL could also easily be deduced by the fact that I mentioned the S2MCheck variable in the topic (which had only been added with the latest commit when I opened this issue). My observations were using the standard profiles and settings (e.g. T19M with 450s duration and a variance of 20%) with the only adjustment being fixed_time=1 because otherwise S2M profiles seem to have very wonky results. I had also confirmed them using other settings though so I didn't put a specific profile in my issue.

I intentionally called my suggested fix a bandaid because it exclusively yielded better results than the current formula (at that point in time) without breaking anything or requiring time to improve the actual formula. (Why? Because the existing formula provided better results at <35% haste and the capped formula provided better results at >35% haste so using a MIN combining both would yield exclusively better results than either.)
However, if you look at how haste scales during high VF stacks, you will most certainly have non-linear interactions with haste since only the Auspicious Spirits part of your insanity generation truly scales linearly. Other abilities eventually have breakpoints depending on the 0.75 seconds minimum GCD in relation to their actual cooldown at the current haste value. That means you might end up having to use a MIN or something equivalent anyway, although likely not for haste in total but for certain parts of the formula.
I'm not going to go deeper into this as the formula is certainly not complete (e.g. no checking for auspicious spirits and crit, shadowy insight and legendary belt, etc.) or universally applicable yet. Having said that, I figured some people might appreciate having an APL that doesn't start giving you completely false haste weightings after around 33% haste and instead gives you conservative ones (by popping S2M as if you didn't have the extra haste) for the time being. I also figured that you might be able to improve the formula in the future knowing that there's an issue occurring at specifically that haste value. Neither seems to be the case so I'll just close this issue.
